From 338b2c16c6b0a76a8b5b2421053ccf027cd73b4f Mon Sep 17 00:00:00 2001
From: kszaq <kszaquitto@gmail.com>
Date: Sun, 25 Sep 2016 03:26:18 +0200
Subject: [PATCH] kernel/power: workaround for Android PM sequence

Suspend sequence in Android kernel is:
1. Suspend -> return 0 (userspace not aware of Android sequence thinks system is already up)
2. Call Early suspend handlers
3. Really suspend

At step 2 userspace is already calling wakeup handlers. To prevent e.g. re-loading modules
before actual suspend and in result preventing it, we need to delay returning a little.
---
 kernel/power/main.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/power/main.c b/kernel/power/main.c
index dd56bc8..6e73ad4 100644
--- a/kernel/power/main.c
+++ b/kernel/power/main.c
@@ -362,6 +362,9 @@ static ssize_t state_store(struct kobject *kobj, struct kobj_attribute *attr,
 			error = 0;
 			request_suspend_state(state);
 		}
+		if (state == PM_SUSPEND_MEM) {
+			msleep_interruptible(5000);
+		}
 #else
 		error = pm_suspend(state);
 #endif
